### PR TITLE
Proxify

### LINKED ISSUE #23: Proxify returned provider
Rather than returning a list of RPCs considered valid based on their `eth_getBlockNumber` response and the burden being on the package user to handle call failures > endpoint reselection via the module, it should be built-in directly.

To do this, the package should return a JS `Proxy` to the provider. This `Proxy` should effectively handle call retries, RPC reselection, error feedback and be configurable.

This means that any read/write op a user makes will now pass through the module (as opposed to only read ops in the linked issue) which is a lot of ground to cover in terms of potential errors.

A proxy user should never need to get involved with errors unless they have caused them more or less and since what they assign as their `"app.provider"` is our `Proxy` they should never need to change provider.

<hr/>

- The proxy should be the default
- Expose both the proxy and a non-proxied version (basically what the package does now)
- configurables: breakpoints, tiered log output, log message constants
- tests should cover read & write ops, as well as any new functionality

<hr/>

The intention is that for a package user everything out of their control is abstracted away, they ask for a provider and it just works unless ofc it's user error.

The expectation is that the user would receive valuable error feedback only if the package is not able to complete the op and this really should only ever happen because of user error.

User-side [Proxy implementation](https://github.com/ubq-testing/pay.ubq.fi/blob/development/static/scripts/rewards/web3/onchain-call-handler.ts) for reference

_original context https://github.com/ubiquity/pay.ubq.fi/issues/235_

##### time: hard to say, <1 day maybe more

### PR DESCRIPTION
**Pull request by Keyrxng**
Resolves #23

Should fix https://github.com/ubiquity/pay.ubq.fi/issues/235

- Added `Proxy` and non-proxy returned providers
- Adopted `Ubiquity/Logger` type logs
- Added tests covering logs and retry logic
- Simplified the structure all into `/types` removing the need for a distinct test esbuild flow
- updated the `yarn build` command which I hope will fix the issue where the NPM push actually has the types included. I thought perhaps the `postbuild` was not executing in the wf? We'll see after this is bumped.

<hr/>

Took a generic approach to errors and have attempted to fill logs with as much detail as available.

Codebases will require an update as there is now a new config option `proxySettings`.

QA:

- I have disabled solvency and `isClaimed` checks for the sake of QA

- The TX I'm sending is just `eth_call, [null,null,null]` which will never succeed

https://github.com/ubiquity/rpc-handler/assets/106303466/04d633ad-aca1-4059-8ddb-dc7442860f22

https://github.com/ubiquity/rpc-handler/assets/106303466/4321adf8-9b8b-4ae6-8bcb-5f1960d384f2

Balance and allowance shown meaning read ops are working as shown with no MM connection, it then succeeds in telling me I've claimed that permit already using `nonceBitmap` 

https://github.com/ubiquity/rpc-handler/assets/106303466/77e0fe19-e173-4715-be31-58cb9a1d7b16

image

### CODE REVIEW

**Comment by Keyrxng** [comment]
Trying to decide what to log and at what tier was difficult because you want signal over noise which is what I've tried to aim for but I feel it's pretty optimal.

I think it's better for the package to throw an error instead of returning one so that regardless of `logTier` they'll know if something is broken.

File: README.md [hunk 1]
```diff
@@ -1,15 +1,17 @@
...
-
-- Returns the lowest latency provider for a given network ID
-- Drops bad endpoints from the list and creates a runtime/local storage cache
-- Can re-test the cached RPCs by calling `handler.getFastestRpcProvider()`
-- Can be used in both the browser and Node.js
-- Fully configurable and extendable
-- Only uses endpoints which Chainlist report as tracking _no_ data (see [`extraRpcs.js`](https://github.com/DefiLlama/chainlist/blob/main/constants/extraRpcs.js))
+## Why use this package?
+
+- **No more slow RPCs**: No more slow RPCs, no more failed requests, no more headaches
```
Comment (molecula451) [inline README.md:5]: hopfully

File: README.md [hunk 2]
```diff
@@ -26,13 +28,22 @@ import { RPCHandler, HandlerConstructorConfig } from "@ubiquity-dao/rpc-handler/
...
 export function useHandler(networkId: number) {
   const config: HandlerConstructorConfig = {
-    networkId: 1,
-    rpcTimeout: 1500,
-    autoStorage: false,
-    cacheRefreshCycles: 10,
-    networkName: null, // the name will be deduced from the networkId, unless using a custom network
-    networkRpcs: null, // same as networkName, but for injecting additional RPCs
-    runtimeRpcs: null, // same as networkRpcs, although these are considered error-free
+    networkId: 1;
```
Comment (molecula451) [inline README.md]: should we default to Ethereum Mainnet? why?

Comment (Keyrxng) [reply README.md]: It's more just a point of reference than a default but I will convert it into a cut and paste default config. I copied and pasted half of a literal config and half of a typed config here I just noticed too.

**Review by 0x4007 — DISMISSED** [review]
It's always a red flag if you have thousands of lines of code changes in a single pull with any sufficiently mature codebase. 

I looked quickly on my phone and it appears that you merged in the rpc manager package source code in here. That's not right. It should be handled in the separate package and imported here as a dependency. 

--

Sorry just realized this isn't pay.ubq.fi!

**Comment by Keyrxng** [comment]
You had me in the first half ngl 😂

The PR is likely < 1000 additions, it's the yarn.lock that's skewing the numbers. GH should allow .lock files to be excluded from addition/deletion count imo there is no value in counting those hashes.

**Comment by molecula451** [comment]
@Keyrxng kindly fix all the run and jests on CI

**Comment by Keyrxng** [comment]
> @Keyrxng kindly fix all the run and jests on CI

I don't think I'm able to I'm getting resource inaccessible errors and I have changed back to the old lockfile. It told me that the bottom line was wrong so I removed it but it is 100% in the dev branch lockfile so I've replaced it and nothing is changing.

Knip CI is also failing for the same reason

image

**Comment by molecula451** [comment]
is the Jest fixing an issue?

**Comment by Keyrxng** [comment]
> is the Jest fixing an issue?

I don't understand the question sorry. The screenshot is just showing that the tests are all passing if CI could get past the env/auth related issue it seems to be having.

this [run](https://github.com/ubiquity/rpc-handler/actions/runs/9572817938/job/26392996659?pr=24) is saying that the last line is wrong (I think), so I changed it and got the below run output.

this [run](https://github.com/ubiquity/rpc-handler/actions/runs/9572651313/job/26392566547) is saying that `resource is inaccessible` and ` Cannot read properties of undefined (reading 'toUpperCase')` in the `yarn install` step

**Comment by molecula451** [comment]
> > is the Jest fixing an issue?
> 
> I don't understand the question sorry. The screenshot is just showing that the tests are all passing if CI could get past the env/auth related issue it seems to be having.
> 
> this [run](https://github.com/ubiquity/rpc-handler/actions/runs/9572817938/job/26392996659?pr=24) is saying that the last line is wrong (I think), so I changed it and got the below run output.
> 
> this [run](https://github.com/ubiquity/rpc-handler/actions/runs/9572651313/job/26392566547) is saying that `resource is inaccessible` and ` Cannot read properties of undefined (reading 'toUpperCase')` in the `yarn install` step

there are opened PRs with CI running, i think we could ignore Knip as i guess there's been unsuccessful times with a proper fix, but why exactly the Jests are failing? & code looks good i don't think the issue comes from development branch, what do you suggest?, @rndquu do you think it's appropiate to merge this with failing CI?

**Comment by Keyrxng** [comment]
- https://github.com/ubiquity/rpc-handler/pull/24/commits/b494243f197b3a563a2abc85669dd3632b7f6283

In this commit I added `util` which is supposed to be a replacement of `node:util` and was required by the logger but I didn't have a lot of success with it so I later removed it. This is the reason for the lockfile being modified in this PR. 

We had this issue with an earlier PR of mine into this repo. I have tried to replace it with the base branch and got those errors above, so I'm out of ideas I think.

<hr/>

- https://github.com/Keyrxng/web3-rpc-racer-1/actions/runs/9581296358/job/26417761931?pr=5

I've opened a PR against my own dev branch and I'm getting the same errors. 

With the added `linkType: hard` it throws telling me that's shouldn't be there (when it's present on the dev branch)

Removing that line then throws the `Cannot read properties of undefined (reading 'toUpperCase')` error. I do not get the resource inaccessible error that I was getting trying to debug the CI runs here.

**Comment by rndquu** [comment]
@Keyrxng 

> Codebases will require an update as there is now a new config option `proxySettings`

You mean this PR introduces breaking changes, right? So if other packages update `rpc-handler` to this version then they will be broken, correct?

**Comment by rndquu** [comment]
@Keyrxng 

Regarding this error:
```
➤ YN0028: -  linkType: hard
➤ YN0028: +  linkType: hard
➤ YN0028: The lockfile would have been modified by this install, which is explicitly forbidden.
```

Try to remove and create a lock file from scratch. Also make sure your local yarn version is v4 as stated [here](https://github.com/ubiquity/rpc-handler/blob/d5b34465f66aea7cfa9c815010d3647a38de0710/package.json#L89).

P.S. The package works fine

**Comment by Keyrxng** [comment]
@rndquu because I have configured the proxy settings by default technically it won't break any project that would bump to the latest version, they'd just use the defaults.

Realised my system config needed some updating/re-arranging shouldn't take too long

**Comment by Keyrxng** [comment]
@gentlementlegen 

> I think this should be solved, because at the moment you just get a warning that some code is running after Jest tear down, but if that code was actually outputting things using console.log Jest would throw an error saying some code is attempting to log after tear down.

I did what you said and logged output during the promises and confirmed requested vs resolved length etc and Jest did not complain so all of them were resolving so by adding `--runInBand` it limits the resources needed to manage tests by running them in sequence

**Comment by Keyrxng** [comment]
@molecula451 all CI passing except Knip which is having a bad day

File: tests/call-handler.test.ts [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+      const blockNumber = await provider.send("eth_blockNumber", ["latest"]);
+      expect(parseInt(blockNumber)).toBeGreaterThan(0);
+    });
+
+    it("should make a successful eth_call", async () => {
+      provider = await handler.getFastestRpcProvider();
+      const response = await provider.send("eth_call", [nonceBitmapData, "latest"]);
+      if (response) {
+        expect(response).toBeDefined();
+        expect(response).toBe("0x0000000000000000000000000000000000000000000000000000000000000000");
```
Comment (molecula451) [inline tests/call-handler.test.ts]: we need to ensure this is 32 bytes

**Comment by Keyrxng** [comment]
@molecula451 I added those explicit return types back because it was silly removing them considering the  uniformity of the rest of the getter functions

@gentlementlegen The open handle warning returned again so I have had to use `isolateModules` alongside `--runInBand` which seems to have resolved it... again, so I'm hoping it stays resolved this time

@rndquu I'm pretty sure the reason why my npm package has types is because I was publishing with an earlier version of Yarn. And with 4.2.2 (might be 4+) you need to explicitly [call](https://yarnpkg.com/migration/guide#explicitly-call-the-pre-and-post-scripts) the pre and post scripts which wasn't happening in the first iteration

File: tests/logs.test.ts [hunk 4]
```diff
@@ -0,0 +1,371 @@
...
+        cleanLogString(`${DEBUG_CALL_NUMBER}4`),
+      ])
+    );
+
+    debugSpy.mockClear();
+    errorSpy.mockClear();
+
+    const res = await provider.send("eth_call", [nonceBitmapData, "latest"]);
+    expect(res).toBeDefined();
+    expect(res).toBe("0x" + "00".repeat(32));
```
Comment (molecula451) [inline tests/logs.test.ts:352]: ha! catch it!

Comment (molecula451) [reply tests/logs.test.ts:352]: i mean the last one was 32 bytes we could have done an assertion, this seems like a innovative way 😄

**Comment by rndquu** [comment]
@Keyrxng Could you merge https://github.com/Keyrxng/web3-rpc-racer-1/pull/6 ?

**Comment by 0x4007** [comment]
You should not force push because then we have to review all the changes in their entirety instead of from the last viewed commit.

**Review by rndquu — APPROVED** [review]
Works fine

**Review by molecula451 — APPROVED** [review]
great improvement in all edges

**Comment by Keyrxng** [comment]
@0x4007 you mentioned to me to avoid force-pushing but you have also mentioned previously about ensuring PRs contain only my commits. 

Maybe I'm missing something, but for it to have only my commits I need to rebase and that requires a force push because it's a different history than my remote. If I merge, I'd have eresdev's commits as well as mine in this PR. If there is a method I'm missing please lmk and I'll look into it.

**Comment by molecula451** [comment]
> @0x4007 you mentioned to me to avoid force-pushing but you have also mentioned previously about ensuring PRs contain only my commits.
> 
> Maybe I'm missing something, but for it to have only my commits I need to rebase and that requires a force push because it's a different history than my remote. If I merge, I'd have eresdev's commits as well as mine in this PR. If there is a method I'm missing please lmk and I'll look into it.

Squash and Merge, remove the comment, leave 1 comment commit, maybe try to do on a test repo

**Comment by Keyrxng** [comment]
> Squash and Merge, remove the comment, leave 1 comment commit, maybe try to do on a test repo

Ty @molecula451, is this what you do? It seems like this would make it appear to reviewers that all the work I've just squashed I authored because it would still be seen in the 'changes made by this pr' but I will give this a try as you suggest

**Comment by molecula451** [comment]
> > Squash and Merge, remove the comment, leave 1 comment commit, maybe try to do on a test repo
> 
> Ty @molecula451, is this what you do? It seems like this would make it appear to reviewers that all the work I've just squashed I authored because it would still be seen in the 'changes made by this pr' but I will give this a try as you suggest

indeed

**Comment by Keyrxng** [comment]
Any objections to merging this folks?

**Review by 0x4007 — APPROVED** [review]
I assume all is well since you know this codebase the best too

**Comment by Keyrxng** [comment]
> I assume all is well since you know this codebase the best too

Great, as it's been a while I'll give it a final look over and merge. Confident but cautious 😂

**Comment by Keyrxng** [comment]
two minor fixes added to round things off:

- removed the `getRPCHandler()` helper function. It's not needed and when testing in a NextJS project it was logging warnings that a critical dependency is an expression (despite it not being called/imported)
- Fixed the `NetworkName` and `NetworkId` autofill in other projects (worked fine in this codebase but not others)

**Pull request merged**